src/core/sock.c

```diff
@@ -1088,3 +1161,24 @@ void nn_sock_stat_increment (struct nn_sock *self, int name, int64_t increment)
             break;
     }
 }
+
+int nn_sock_hold (struct nn_sock *self)
+{
+    switch (self->state) {
+    case NN_SOCK_STATE_ACTIVE:
+    case NN_SOCK_STATE_INIT:
+        self->holds++;
+        return 0;
+    case NN_SOCK_STATE_ZOMBIE:
+        return -ETERM;
+    }
```

Comment:
I have different opinions about assert()  -- its useful for documenting conditions that the programmer is certain of; external errnos don't fall into that category.

EBADF is used only *here* because its the most appropriate semantic to use.  We do have EFSM already which I could use for the default case, if it really bothers you.  (My feeling is that any other states that may get added should be treated as not-ready in this case, i.e. closing or not initialized yet, therefore EBADF.)

I don't want to create a new magic errno.  EBADF has particular semantics here -- the file handle (socket descriptor) is invalid.  (Usually because it is closed or somesuch.)

nn_assert() needs to *always* be a panic stop.  Just logging & continuing makes it something much different than an assertion.  The Gnome people went that way years ago, and as a result gnome programs spew assertions like crazy -- because they assert things that clearly are false or not assertions, and yet the program continues onward.  That's abuse of assertions.

assert() is like crashing a kernel.  Its to be used only when you've got some fundamental assumption, and if that assumption is broken then the very worst thing you can do is try to continue, because it means you have no idea what the state of the program is.

If asserts are firing and causing problems, then those are critical bugs that need to be fixed.  In many case we just need to remove the assertions.  (For example, handling of errnos -- instead of an assertion we should inject correct handling to deal with unexpected system errors.)

Anyway, this is way out of scope for this change set.